django/contrib/postgres/operations.py

```diff
@@ -164,3 +164,65 @@ def database_backwards(self, app_label, schema_editor, from_state, to_state):
             to_model_state = to_state.models[app_label, self.model_name_lower]
             index = to_model_state.get_index_by_name(self.name)
             schema_editor.add_index(model, index, concurrently=True)
+
+
+class CreateCollation(Operation):
+    reversible = True
+
+    def __init__(self, name, *, locale=None, provider=None, deterministic=True):
+        self.name = name
+        self.locale = locale
+        self.provider = provider
+        self.deterministic = deterministic
+
+    def state_forwards(self, app_label, state):
+        pass
+
+    def database_forwards(self, app_label, schema_editor, from_state, to_state):
+        if (
+            schema_editor.connection.vendor != 'postgresql' or
+            not router.allow_migrate(schema_editor.connection.alias, app_label)
+        ):
+            return
+        if (
+            self.deterministic is False and
+            not schema_editor.connection.features.supports_non_deterministic_collations
+        ):
+            raise NotSupportedError('Non-deterministic collations require PostgreSQL 12 or above')
+        if not self.collation_exists(schema_editor, self.name):
+            args = []
+            if self.locale:
+                args.append(f'locale = {schema_editor.quote_name(self.locale)}')
+            if self.provider:
+                args.append(f"provider = {schema_editor.quote_name(self.provider)}")
+            # Before PostgreSQL 12, the deterministic parameter doesn't exist,
+            # so only set it if it's False.
+            if self.deterministic is False:
+                args.append('deterministic = false')
+            sql_args = ', '.join(args)
+            schema_editor.execute(
+                f'CREATE COLLATION IF NOT EXISTS {schema_editor.quote_name(self.name)} ({sql_args})'
```

Comment:
Collations are not the same as extensions, they are not determine by names. IMO we shouldn't use `IF (NOT) EXISTS` syntax but fail loudly instead. It can create a tricky issues when we will omit creating collations just because the collation with the same name already exists.

Reply:
Agreed. In particular, the PostgreSQL [documentation](https://www.postgresql.org/docs/current/sql-createcollation.html#id-1.9.3.59.6) states:

> Note that there is no guarantee that the existing collation is anything like the one that would have been created.

Reply:
Makes sense, I've changed it.

Reply:
We should also remove `self.collation_exists(schema_editor, self.name)` checks.